pkg/envoy/server.go

```diff
@@ -237,7 +237,25 @@ func (s *XDSServer) AddListener(name string, endpointPolicyName string, port uin
 
 	listenerConf.FilterChains[0].Filters[1].Config.Fields["http_filters"].GetListValue().Values[0].GetStructValue().Fields["config"].GetStructValue().Fields["policy_name"] = &structpb.Value{Kind: &structpb.Value_StringValue{StringValue: endpointPolicyName}}
 
-	s.listenerMutator.Upsert(ListenerTypeURL, name, listenerConf, []string{"127.0.0.1"}, wg.AddCompletion())
+	retries := 0
+	maxRetries := 10
+	var comp *completion.Completion
+	comp = wg.AddCompletionWithCallbacks(nil, func() {
+		if retries < maxRetries {
+			retries++
+			// Upsert cannot be called from the callback due to a locks being held
+			go func() {
+				// Upsert a new version of the Listener
+				// TODO: Try to recover by reallocating the proxy port
+				s.listenerMutator.Upsert(ListenerTypeURL, name, listenerConf, []string{"127.0.0.1"}, comp)
```

Comment:
Have you tested that this works?
If I read this correctly, `comp` is `nil` by the time this closure is created. This is definitely not what you want.
I think you need to create a new `Completion` here.

You have a circular dependency problem.
You can't create the completion without the closure, and you can't create the closure without the completion.
:/

Reply:
Not tested yet, but closure should not cache variable values. `comp` is a pointer and it is set to non-nil value before the closure is used. At least in the tests incrementing a counter in the completion callback works.

Reply:
I added a server_e2e_test with this same pattern and it works!